File: java/matching-conformance/src/main/java/io/github/giovanicaprison/matching/conformance/FixtureParser.java
package io.github.giovanicaprison.matching.conformance;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * Reads the fixture format described in {@code TESTING.md}.
 *
 * <p>Whitespace separated words, one line per command or expected event, so a runner in any
 * language is a small amount of code. That matters more than terseness here: the fixtures are the
 * contract between a Java engine and a C++ one, and a format that is awkward to parse invites two
 * readings of it.
 *
 * <p>A malformed line fails immediately and names the file and the line. A fixture that parses
 * loosely would produce a comparison failure somewhere else entirely.
 */
public final class FixtureParser {

  /** The least number of words each command needs, before any optional ones. */
  private static final Map<Directive, Integer> ARITY = arity();

  private FixtureParser() {}

  public static Fixture parse(final Path file) {
    try {
      return parse(file.getFileName().toString(), Files.readString(file));
    } catch (final IOException e) {
      throw new UncheckedIOException("cannot read fixture " + file, e);
    }
  }

  public static Fixture parse(final String name, final String content) {
    final List<Fixture.Element> elements = new ArrayList<>();
    String title = "";
    int number = 0;
    for (final String raw : content.lines().toList()) {
      number++;
      final String line = raw.strip();
      if (line.startsWith("#")) {
        if (title.isEmpty()) {
          title = line.substring(1).strip();
        }
        continue;
      }
      if (line.isEmpty()) {
        continue;
      }
      elements.add(element(name, number, line));
    }
    final Fixture fixture = new Fixture(name, title, List.copyOf(elements));
    requireInstrumentFirst(fixture);
    return fixture;
  }

  private static Fixture.Element element(final String name, final int number, final String line) {
    final List<String> words = Arrays.asList(line.split("\\s+"));
    final String first = words.getFirst();
    final Directive directive = lookup(Directive.class, first);
    if (directive != null) {
      final List<String> arguments = words.subList(1, words.size());
      if (arguments.size() < ARITY.get(directive)) {
        throw new MalformedFixture(
            name, number, directive + " needs at least " + ARITY.get(directive) + " words");
      }
      return new Fixture.Command(line, number, directive, List.copyOf(arguments));
    }
    if (lookup(Verb.class, first) != null) {
      return new Fixture.Expected(line, number);
    }
    throw new MalformedFixture(name, number, first + " is neither a directive nor an output verb");
  }

  private static void requireInstrumentFirst(final Fixture fixture) {
    final List<Fixture.Command> commands = fixture.commands();
    if (commands.isEmpty() || commands.getFirst().directive() != Directive.INSTRUMENT) {
      throw new MalformedFixture(fixture.name(), 1, "the first command must be INSTRUMENT");
    }
    if (commands.stream()
        .skip(1)
        .anyMatch(command -> command.directive() == Directive.INSTRUMENT)) {
      throw new MalformedFixture(
          fixture.name(), 1, "an instrument is defined once for the life of an engine");
    }
  }

  private static <E extends Enum<E>> E lookup(final Class<E> type, final String word) {
    for (final E value : type.getEnumConstants()) {
      if (value.name().equals(word)) {
        return value;
      }
    }
    return null;
  }

  private static Map<Directive, Integer> arity() {
    final Map<Directive, Integer> arity = new EnumMap<>(Directive.class);
    arity.put(Directive.INSTRUMENT, 1);
    arity.put(Directive.SESSION, 1);
    arity.put(Directive.NEW, 5);
    arity.put(Directive.CANCEL, 1);
    arity.put(Directive.REPLACE, 3);
    arity.put(Directive.MASSCANCEL, 1);
    return Map.copyOf(arity);
  }

  /** A fixture the runner refuses to guess at. */
  public static final class MalformedFixture extends RuntimeException {

    MalformedFixture(final String name, final int line, final String problem) {
      super(name + ":" + line + " " + problem);
    }
  }
}

File: cpp/conformance/fixture_parser.cpp
#include "conformance/fixture_parser.hpp"

#include <fstream>
#include <sstream>

namespace io::github::giovanicaprison::matching::conformance {

namespace {

// The least number of words each command needs, before any optional ones.
int minimumArity(const Directive directive) {
  switch (directive) {
    case Directive::INSTRUMENT:
    case Directive::SESSION:
      return 1;
    case Directive::NEW:
      return 5;
    case Directive::CANCEL:
      return 1;
    case Directive::REPLACE:
      return 3;
    case Directive::MASSCANCEL:
      return 1;
  }
  return 0;
}

const char* directiveName(const Directive directive) {
  switch (directive) {
    case Directive::INSTRUMENT:
      return "INSTRUMENT";
    case Directive::SESSION:
      return "SESSION";
    case Directive::NEW:
      return "NEW";
    case Directive::CANCEL:
      return "CANCEL";
    case Directive::REPLACE:
      return "REPLACE";
    case Directive::MASSCANCEL:
      return "MASSCANCEL";
  }
  return "?";
}

std::string stripped(const std::string& raw) {
  const auto first = raw.find_first_not_of(" \t\r");
  if (first == std::string::npos) {
    return "";
  }
  const auto last = raw.find_last_not_of(" \t\r");
  return raw.substr(first, last - first + 1);
}

std::vector<std::string> words(const std::string& line) {
  std::vector<std::string> found;
  std::istringstream stream(line);
  std::string word;
  while (stream >> word) {
    found.push_back(word);
  }
  return found;
}

Element element(const std::string& name, const int number, const std::string& line) {
  const std::vector<std::string> parts = words(line);
  const std::string& first = parts.front();
  if (const auto directive = directiveOf(first)) {
    const std::vector<std::string> arguments(parts.begin() + 1, parts.end());
    const int arity = minimumArity(*directive);
    if (std::cmp_less(arguments.size(), arity)) {
      throw MalformedFixture(name, number,
                             std::string(directiveName(*directive)) + " needs at least " +
                                 std::to_string(arity) + " words");
    }
    return Command{line, number, *directive, arguments};
  }
  if (verbOf(first)) {
    return Expected{line, number};
  }
  throw MalformedFixture(name, number, first + " is neither a directive nor an output verb");
}

void requireInstrumentFirst(const Fixture& fixture) {
  const std::vector<Command> commands = fixture.commands();
  if (commands.empty() || commands.front().directive != Directive::INSTRUMENT) {
    throw MalformedFixture(fixture.name, 1, "the first command must be INSTRUMENT");
  }
  for (std::size_t at = 1; at < commands.size(); at++) {
    if (commands[at].directive == Directive::INSTRUMENT) {
      throw MalformedFixture(fixture.name, 1,
                             "an instrument is defined once for the life of an engine");
    }
  }
}

}  // namespace

Fixture parseFixture(const std::string& name, const std::string& content) {
  Fixture fixture;
  fixture.name = name;
  int number = 0;
  std::istringstream lines(content);
  std::string raw;
  while (std::getline(lines, raw)) {
    number++;
    const std::string line = stripped(raw);
    if (!line.empty() && line.front() == '#') {
      if (fixture.title.empty()) {
        fixture.title = stripped(line.substr(1));
      }
      continue;
    }
    if (line.empty()) {
      continue;
    }
    fixture.elements.push_back(element(name, number, line));
  }
  requireInstrumentFirst(fixture);
  return fixture;
}

Fixture parseFixture(const std::filesystem::path& file) {
  std::ifstream stream(file);
  if (!stream) {
    throw std::runtime_error("cannot read fixture " + file.string());
  }
  std::ostringstream content;
  content << stream.rdbuf();
  return parseFixture(file.filename().string(), content.str());
}

}  // namespace io::github::giovanicaprison::matching::conformance
